### DISCUSSION TITLE
Using PocketBase Auth in Next js 13 App Directory.

### DISCUSSION THREAD

**Discussion by ericbaek** [pocketbase/pocketbase#2846] [Q&A] [answered]
## found the answer.
- https://github.com/tigawanna/next-pocketbase-demo
- https://github.com/tsensei/nextjs-pocketbase-starter-template

I'm attempting to create an admin-only page using PocketBase Auth in Next.js 13 (in the App directory). With many changes in this version, sometimes the documentation falls short in helping me 😢. Packages like next-auth offer {getServerSession} to simplify things, but it seems there's not much I can do with PocketBase, or perhaps I just can't find the necessary resources.

I'm currently encountering a Hydration error because the initial UI doesn't match what was rendered on the server. This is likely because the HTML changes after the user loads the website. Is there a way to ensure that the UI from the server-side pre-rendering matches that of the user's machine? Specifically, how can I synchronize the authentication state with the server-side?

At present, when a user is logged in and then refreshes the page, the initial login form screen briefly appears before updating to 'Welcome Admin' along with a hydration error. I would like to prevent my website from rendering the login form if the user is already logged in (or encountering any errors). Additionally, if I want to restrict access to certain pages for non-admin users, how can I utilize middleware to enforce this using PocketBase Auth?

```
import PocketBase from 'pocketbase';

const pb = new PocketBase('URL');
export default pb;
```

```
'use client'

import { Admin, Record } from "pocketbase";
import pb from "../utils/pocketbase";
import { useForm } from "react-hook-form";
import { useState } from "react";

export default function Page() {
    const { register, handleSubmit } = useForm();
    const isLoggedOn = (pb.authStore.model instanceof Admin);
    const [isLoading, setLoading] = useState(false);

    async function logon(data: any) {
        setLoading(true);
        try {
            const authData = await pb.admins.authWithPassword(data.email, data.password);
            setLoading(false);
        } catch (e) {
            alert(e);
            setLoading(false);
        }
    }

    function logout() {
        setLoading(true);
        pb.authStore.clear();
        setLoading(false);
    }

    if (pb.authStore.model instanceof Admin) {
        console.log("admin")
    } else if (pb.authStore.model instanceof Record) {
        console.log("auth record")
    } else {
        console.log("Guest")
    }

    if (isLoggedOn) {
        return (
            <>
                <h1>Welcome Admin.</h1>
                <button onClick={logout}>Log out</button>
            </>
        )
    }

    return (
        <>
            <form onSubmit={handleSubmit(logon)}>
                <h1>Please log on.</h1>
                <input type="text" placeholder="ID" {...register("email")} />
                <input type="password" placeholder="Password" {...register("password")} />
                <button type="submit" disabled={isLoading}>Log on</button>
            </form>
        </>
    )
}
```

**Comment by ganigeorgiev** [comment]
As mentioned in https://github.com/pocketbase/pocketbase/discussions/2746#discussioncomment-6349769, do you need pre-rendering for this page in the first place? I'm not really sure that I understand what is the expected outcome of pre-rendering it. You can't have dynamic/conditional processing during the pre-rendering phase (_unless I'm misunderstanding how this works in Next.js_).

The above page will always differ when executed in a browser vs server context because the browser uses localStorage and if you have previously authenticated it will try to reuse the token, aka. after login as admin, the `isLoggedOn` check will be true on the next refresh when executed inside the browser. In the linked discussion there is an example how to "disable" the `pb.authStore` `localStorage` behavior, but again is this what you really want?

**Reply by ericbaek** [reply]
I don't need pre-rendering for this page. but didn't realize that was an option to be turned off.
I thought by stating `use client` it becomes CSR, therefore does not pre-render.

**Reply by ganigeorgiev** [reply]
Also note that if `import pb from "../utils/pocketbase";` returns a global `PocketBase` JS SDK instance, this will not work in SSR context, because requests from different users will write over the same `pb.authStore` state. In this case you need to initialize a new `PocketBase` instance for each request and store it as request local/context prop value.

**Comment by tigawanna** [comment, marked as answer]
Did a few experiments and this worked for me , too long to post it all here [readme link](https://github.com/tigawanna/next-pocketbase-demo#readme)

**Comment by tsensei** [comment]
I have written a solution and some helper functions to help with cookie based auth, please check out here : https://github.com/pocketbase/js-sdk/issues/69#issuecomment-1840433737

**Comment by 0xsarwagya** [comment]
A Possible Solution May Be Using NextAuth.js With Pocketbase Adapter. Our Team Has Created An Adapter Recently Since We Are Using Pocketbase Now.

It's Still Under Active Development and Contributions are Welcome

[You Can look at The Github Repo Here](https://github.com/Rebackk-Team/pocketbase-adapter)
[Also The NPM Package Here](https://www.npmjs.com/package/@rebackk/pocketbase-adapter)

**Reply by mbecker** [reply]
Hi, I've looked at your adapter. You are basically adding complete new tables in pocktebase to fulfil the nextauth documentation for db adapters. That means the authentication is not done by pocketbase and user's authentication toke is returned. The token is necessary to request records which are protected by an API rule.

**Comment by mbecker** [comment]
Hi,

I haven't looked at the code above. But I have used nextjs, nextauth and pocketbase in a project for a while.

I've created a public github repo for it: [https://github.com/mbecker/nextjs-nextauth-pocketbase](https://github.com/mbecker/nextjs-nextauth-pocketbase)

A live demo is here: [https://nextjs-nextauth-pocketbase.vercel.app/](https://nextjs-nextauth-pocketbase.vercel.app/)
(nextjs deployed to vercel, pocketbase hosted at pockethost.io)

My requirements are that
- OAuth providers are configured in  pocketbase 
- pocketbase handles the authentication flow at the OAuth provider
- pocketbase returns a user token (the user's token is importnat to request protected collections)
- nextjs and nextauth techstack (which means use nextauth for frontend authentication and leveraging all nextauth features like `useSession`)
- nextauth used in SSR and client

The idea is more or less documented at the official pocketbase documentation: [OAuth 2.0 integration - Manual code exchange](https://pocketbase.io/docs/authentication/)

This means:
- At the signin page we are requesting all configured oauth providers and displaying it as buttons
- The user clicks a signin/oauth button with the configured redirect uri and the server actions creates a cookie with the authprovider data
- The nextauth custom oauth provider integration handles the redirect from the oauth provider by getting the authentication code from the url and then authenticates at pocketbase with the code + authprovider data from the cookie
- pocketbase authenticates the user and returns an authentication token

I'm trying to start a new discussion to have a central to discuss the idea of my repo this evening / tomorrow. Additionally, issues in the repo would be nice!

Hope this helps!
